arwen/contexts/async.go

```diff
@@ -263,66 +351,44 @@ func (context *asyncContext) Execute() error {
 	return nil
 }
 
-// PrepareLegacyAsyncCall builds an AsyncCall struct from its arguments, sets it as
-// the default async call and informs Wasmer to stop contract execution with BreakpointAsyncCall
-func (context *asyncContext) PrepareLegacyAsyncCall(address []byte, data []byte, value []byte) error {
-	legacyGroupID := arwen.LegacyAsyncCallGroupID
-	legacyCallback := []byte(arwen.CallbackFunctionName)
-
-	_, exists := context.GetCallGroup(legacyGroupID)
-	if exists {
-		return arwen.ErrOnlyOneLegacyAsyncCallAllowed
-	}
-
-	err := context.CreateAndAddCall(legacyGroupID,
-		address,
-		data,
-		value,
-		legacyCallback,
-		legacyCallback,
-		math.MaxUint64,
-	)
-	if err != nil {
-		return err
+// executeAsyncCall executes asynchronous cross-shard calls to built-in
+// functions and contracts
+func (context *asyncContext) executeAsyncCall(asyncCall *arwen.AsyncCall) error {
+	if asyncCall.ExecutionMode == arwen.AsyncUnknown {
+		return context.sendAsyncCallCrossShard(asyncCall)
 	}
 
-	context.host.Runtime().SetRuntimeBreakpointValue(arwen.BreakpointAsyncCall)
-
-	return nil
-}
+	if asyncCall.ExecutionMode == arwen.AsyncBuiltinFunc {
+		// Built-in functions will handle cross-shard calls themselves, by
```

Comment:
make a separate func for all the code inside this if.

Reply:
Done.